In [15]:
import pandas as pd
import numpy as np
from pathlib import Path
from copy import copy

In [4]:
PATH_RAW = Path('../data/raw')

# Loading Data

In [5]:
day_3_data = pd.read_csv(PATH_RAW / "day_3.txt", header=None).values

# Step 1

--- Day 3: Toboggan Trajectory ---

With the toboggan login problems resolved, you set off toward the airport. While travel by toboggan might be easy, it's certainly not safe: there's very minimal steering and the area is covered in trees. You'll need to see which angles will take you near the fewest trees.

Due to the local geology, trees in this area only grow on exact integer coordinates in a grid. You make a map (your puzzle input) of the open squares (.) and trees (#) you can see. For example:

..##.......
#...#...#..
.#....#..#.
..#.#...#.#
.#...##..#.
..#.##.....
.#.#.#....#
.#........#
#.##...#...
#...##....#
.#..#...#.#
These aren't the only trees, though; due to something you read about once involving arboreal genetics and biome stability, the same pattern repeats to the right many times:

..##.........##.........##.........##.........##.........##.......  --->
#...#...#..#...#...#..#...#...#..#...#...#..#...#...#..#...#...#..
.#....#..#..#....#..#..#....#..#..#....#..#..#....#..#..#....#..#.
..#.#...#.#..#.#...#.#..#.#...#.#..#.#...#.#..#.#...#.#..#.#...#.#
.#...##..#..#...##..#..#...##..#..#...##..#..#...##..#..#...##..#.
..#.##.......#.##.......#.##.......#.##.......#.##.......#.##.....  --->
.#.#.#....#.#.#.#....#.#.#.#....#.#.#.#....#.#.#.#....#.#.#.#....#
.#........#.#........#.#........#.#........#.#........#.#........#
#.##...#...#.##...#...#.##...#...#.##...#...#.##...#...#.##...#...
#...##....##...##....##...##....##...##....##...##....##...##....#
.#..#...#.#.#..#...#.#.#..#...#.#.#..#...#.#.#..#...#.#.#..#...#.#  --->
You start on the open square (.) in the top-left corner and need to reach the bottom (below the bottom-most row on your map).

The toboggan can only follow a few specific slopes (you opted for a cheaper model that prefers rational numbers); start by counting all the trees you would encounter for the slope right 3, down 1:

From your starting position at the top-left, check the position that is right 3 and down 1. Then, check the position that is right 3 and down 1 from there, and so on until you go past the bottom of the map.

The locations you'd check in the above example are marked here with O where there was an open square and X where there was a tree:

..##.........##.........##.........##.........##.........##.......  --->
#..O#...#..#...#...#..#...#...#..#...#...#..#...#...#..#...#...#..
.#....X..#..#....#..#..#....#..#..#....#..#..#....#..#..#....#..#.
..#.#...#O#..#.#...#.#..#.#...#.#..#.#...#.#..#.#...#.#..#.#...#.#
.#...##..#..X...##..#..#...##..#..#...##..#..#...##..#..#...##..#.
..#.##.......#.X#.......#.##.......#.##.......#.##.......#.##.....  --->
.#.#.#....#.#.#.#.O..#.#.#.#....#.#.#.#....#.#.#.#....#.#.#.#....#
.#........#.#........X.#........#.#........#.#........#.#........#
#.##...#...#.##...#...#.X#...#...#.##...#...#.##...#...#.##...#...
#...##....##...##....##...#X....##...##....##...##....##...##....#
.#..#...#.#.#..#...#.#.#..#...X.#.#..#...#.#.#..#...#.#.#..#...#.#  --->
In this example, traversing the map using this slope would cause you to encounter 7 trees.

Starting at the top-left corner of your map and following a slope of right 3 and down 1, how many trees would you encounter?

In [118]:
data_map = day_3_data
length_j = len(data_map[0][0])
print_result = []
result = []

for row_count, i in enumerate(range(40)):
    j = (row_count * 3) % length_j

    value = data_map[i][0][j]
    
    if value=='.':
        print_result.append(['.'*(j) + 'O' + '.'*(length_j-j-1)])
    else:
        print_result.append(['.'*(j) + 'X' + '.'*(length_j-j-1)])
    result.append(value=='#')

print(np.sum(result))
print_result

34


[['O..............................'],
 ['...X...........................'],
 ['......X........................'],
 ['.........O.....................'],
 ['............O..................'],
 ['...............X...............'],
 ['..................X............'],
 ['.....................X.........'],
 ['........................X......'],
 ['...........................X...'],
 ['..............................X'],
 ['..X............................'],
 ['.....X.........................'],
 ['........X......................'],
 ['...........X...................'],
 ['..............X................'],
 ['.................X.............'],
 ['....................X..........'],
 ['.......................X.......'],
 ['..........................X....'],
 ['.............................O.'],
 ['.X.............................'],
 ['....X..........................'],
 ['.......X.......................'],
 ['..........X....................'],
 ['.............X.................'],
 ['.........

In [119]:
def trees_encoutred(data_map: [str]) -> int:
    length_j = len(data_map[0][0])
    print_result = []
    result = []

    for row_count, i in enumerate(range(len(data_map))):
        j = (row_count * 3) % length_j

        value = data_map[i][0][j]
        result.append(value=='#')
    
    return np.sum(result)

In [120]:
trees_encoutred(data_map=day_3_data)

278

# Step 2


Your puzzle answer was 278.

The first half of this puzzle is complete! It provides one gold star: *

--- Part Two ---

Time to check the rest of the slopes - you need to minimize the probability of a sudden arboreal stop, after all.

Determine the number of trees you would encounter if, for each of the following slopes, you start at the top-left corner and traverse the map all the way to the bottom:

Right 1, down 1.
Right 3, down 1. (This is the slope you already checked.)
Right 5, down 1.
Right 7, down 1.
Right 1, down 2.
In the above example, these slopes would find 2, 7, 3, 4, and 2 tree(s) respectively; multiplied together, these produce the answer 336.

What do you get if you multiply together the number of trees encountered on each of the listed slopes?

In [125]:
def angle_trees_encoutred(data_map: [str], config=[int, int]) -> int:
    length_j = len(data_map[0][0])
    result = []
    print_result = []

    for row_count, i in enumerate(range(0, len(data_map), config[1])):
        j = (row_count * config[0]) % length_j

        value = data_map[i][0][j]
        result.append(value=='#')

    return np.sum(result)

In [126]:
configs = [[1, 1], [3, 1], [5, 1], [7, 1], [1, 2]]
result = []

for config in configs:
    result.append(angle_trees_encoutred(data_map=day_3_data, 
                                        config=config))

result

[90, 278, 88, 98, 45]

In [127]:
product_of_encountred_trees = np.prod(result)
product_of_encountred_trees

9709761600